File: tenet-0/shared/bus-go/constitution.go
package bus

import (
	"context"
	"fmt"
	"time"
)

// Constitution returns the Constitution view of this Bus.
func (b *Bus) Constitution() *Constitution {
	return &Constitution{bus: b}
}

// Constitution exposes the active versioned constitution and a Watch
// callback for hot-reloading agent prompts when the constitution is bumped.
type Constitution struct {
	bus *Bus
}

// LoadedConstitution is the active constitution as seen by an agent.
type LoadedConstitution struct {
	VersionID int64
	ProseText string
	RulesYAML string
	Rules     []ConstitutionRule
}

// ConstitutionRule is one parsed rule from the active constitution.
type ConstitutionRule struct {
	RuleID               string
	EventTypePattern     string
	RequiresApprovalMode string // "per_action", "blanket_category", "none", or ""
	ApprovalCategory     string
}

// Load returns the currently active constitution. Agents call this at startup
// (and on each Watch callback) to embed the constitution into their system
// prompt.
func (c *Constitution) Load(ctx context.Context) (*LoadedConstitution, error) {
	var loaded LoadedConstitution
	err := c.bus.pool.QueryRow(ctx,
		`SELECT version_id, prose_text, rules_yaml
		   FROM constitution_versions WHERE is_active LIMIT 1`,
	).Scan(&loaded.VersionID, &loaded.ProseText, &loaded.RulesYAML)
	if err != nil {
		return nil, fmt.Errorf("constitution: load: %w", err)
	}

	rows, err := c.bus.pool.Query(ctx,
		`SELECT rule_id, event_type_pattern,
		        COALESCE(requires_approval_mode, ''),
		        COALESCE(approval_category, '')
		   FROM constitution_rules
		  WHERE constitution_version_id = $1
		  ORDER BY id`,
		loaded.VersionID,
	)
	if err != nil {
		return nil, fmt.Errorf("constitution: load rules: %w", err)
	}
	defer rows.Close()

	for rows.Next() {
		var r ConstitutionRule
		if err := rows.Scan(&r.RuleID, &r.EventTypePattern, &r.RequiresApprovalMode, &r.ApprovalCategory); err != nil {
			return nil, fmt.Errorf("constitution: scan rule: %w", err)
		}
		loaded.Rules = append(loaded.Rules, r)
	}
	return &loaded, rows.Err()
}

// CurrentVersion returns the active version_id without loading prose/rules.
// Cheap; safe to call from a polling loop.
func (c *Constitution) CurrentVersion(ctx context.Context) (int64, error) {
	var v int64
	err := c.bus.pool.QueryRow(ctx,
		`SELECT version_id FROM constitution_versions WHERE is_active LIMIT 1`,
	).Scan(&v)
	if err != nil {
		return 0, fmt.Errorf("constitution: current version: %w", err)
	}
	return v, nil
}

// Watch polls CurrentVersion at the given interval and invokes onChange with
// the new version whenever it changes. Returns a stop function the caller
// invokes to cancel the watcher; the watcher also stops when ctx is done.
//
// onChange is invoked synchronously inside the polling goroutine — a slow
// callback delays detection of the next change. Offload to a goroutine if
// the work may exceed interval.
//
// Each agent's loop should call Load() inside onChange to refresh its prompt
// at the next task boundary.
//
// To prevent runaway polling from misconfigured callers (test intervals
// leaking into production), the minimum effective interval is 1 second.
func (c *Constitution) Watch(ctx context.Context, interval time.Duration, onChange func(newVersion int64)) (stop func(), err error) {
	if interval <= 0 {
		return nil, fmt.Errorf("constitution: watch interval must be positive")
	}
	const minInterval = 1 * time.Second
	if interval < minInterval {
		c.bus.logger.Warn("constitution: watch interval below minimum, clamping",
			"requested", interval, "effective", minInterval)
		interval = minInterval
	}

	current, err := c.CurrentVersion(ctx)
	if err != nil {
		return nil, err
	}

	watchCtx, cancel := context.WithCancel(ctx)
	done := make(chan struct{})

	go func() {
		defer close(done)
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-watchCtx.Done():
				return
			case <-ticker.C:
				v, err := c.CurrentVersion(watchCtx)
				if err != nil {
					if watchCtx.Err() != nil {
						return
					}
					c.bus.logger.Warn("constitution: watch poll failed", "error", err)
					continue
				}
				if v != current {
					current = v
					onChange(v)
				}
			}
		}
	}()

	return func() { cancel(); <-done }, nil
}

File: tenet-0/shared/bus-ts/src/constitution.ts
import type { Bus } from "./bus.js";

export interface ConstitutionRule {
  ruleId: string;
  eventTypePattern: string;
  requiresApprovalMode: string; // "per_action" | "blanket_category" | "none" | ""
  approvalCategory: string;
}

export interface LoadedConstitution {
  versionId: number;
  proseText: string;
  rulesYaml: string;
  rules: ConstitutionRule[];
}

const MIN_WATCH_INTERVAL_MS = 1000;

// Constitution exposes the active versioned constitution and a watch helper
// for hot-reloading agent prompts when the constitution is bumped.
export class Constitution {
  private readonly bus: Bus;

  constructor(bus: Bus) {
    this.bus = bus;
  }

  async load(): Promise<LoadedConstitution> {
    const { rows: versionRows } = await this.bus.pool.query(
      `SELECT version_id, prose_text, rules_yaml
         FROM constitution_versions WHERE is_active LIMIT 1`,
    );
    const v = versionRows[0];
    if (!v) throw new Error("constitution: no active version");

    const { rows: ruleRows } = await this.bus.pool.query(
      `SELECT rule_id, event_type_pattern,
              COALESCE(requires_approval_mode, '') AS requires_approval_mode,
              COALESCE(approval_category, '') AS approval_category
         FROM constitution_rules
        WHERE constitution_version_id = $1
        ORDER BY id`,
      [v.version_id],
    );

    return {
      versionId: Number(v.version_id),
      proseText: v.prose_text,
      rulesYaml: v.rules_yaml,
      rules: ruleRows.map((r) => ({
        ruleId: r.rule_id,
        eventTypePattern: r.event_type_pattern,
        requiresApprovalMode: r.requires_approval_mode,
        approvalCategory: r.approval_category,
      })),
    };
  }

  async currentVersion(): Promise<number> {
    const { rows } = await this.bus.pool.query(
      `SELECT version_id FROM constitution_versions WHERE is_active LIMIT 1`,
    );
    if (!rows[0]) throw new Error("constitution: no active version");
    return Number(rows[0].version_id);
  }

  // watch polls currentVersion at the given interval and invokes onChange when
  // the version changes. Returns a stop() function. The minimum effective
  // interval is 1 second — smaller values are clamped.
  async watch(
    intervalMs: number,
    onChange: (newVersion: number) => void | Promise<void>,
  ): Promise<() => Promise<void>> {
    if (intervalMs <= 0) {
      throw new Error("constitution: watch interval must be positive");
    }
    const effective = Math.max(intervalMs, MIN_WATCH_INTERVAL_MS);
    if (effective !== intervalMs) {
      console.warn(
        `constitution: watch interval ${intervalMs}ms below minimum, clamped to ${effective}ms`,
      );
    }

    let current = await this.currentVersion();
    let stopped = false;
    let timer: NodeJS.Timeout | null = null;
    let pending: Promise<void> | null = null;

    const tick = async (): Promise<void> => {
      if (stopped) return;
      try {
        const v = await this.currentVersion();
        if (v !== current) {
          current = v;
          await onChange(v);
        }
      } catch (err) {
        if (!stopped) {
          console.warn("constitution: watch poll failed:", err);
        }
      }
    };

    timer = setInterval(() => {
      if (pending) return; // Skip tick if the last one is still running.
      pending = tick().finally(() => {
        pending = null;
      });
    }, effective);

    return async () => {
      stopped = true;
      if (timer) clearInterval(timer);
      if (pending) await pending;
    };
  }
}
